sycl/plugins/unified_runtime/ur/adapters/hip/usm.cpp

```diff
@@ -13,85 +13,86 @@
 #include "context.hpp"
 #include "device.hpp"
 #include "platform.hpp"
+#include "usm.hpp"
 
 /// USM: Implements USM Host allocations using HIP Pinned Memory
-UR_APIEXPORT ur_result_t UR_APICALL urUSMHostAlloc(
-    ur_context_handle_t hContext, const ur_usm_desc_t *pUSMDesc,
-    [[maybe_unused]] ur_usm_pool_handle_t pool, size_t size, void **ppMem) {
-
-  UR_ASSERT(!pUSMDesc || (pUSMDesc->align == 0 ||
-                          ((pUSMDesc->align & (pUSMDesc->align - 1)) == 0)),
+UR_APIEXPORT ur_result_t UR_APICALL
+urUSMHostAlloc(ur_context_handle_t hContext, const ur_usm_desc_t *pUSMDesc,
+               ur_usm_pool_handle_t hPool, size_t size, void **ppMem) {
+  auto alignment = pUSMDesc ? pUSMDesc->align : 0u;
+  UR_ASSERT(!pUSMDesc ||
+                (alignment == 0 || ((alignment & (alignment - 1)) == 0)),
             UR_RESULT_ERROR_INVALID_VALUE);
+  UR_ASSERT(size < hContext->DeviceId->getMaxAllocSize(),
+            UR_RESULT_ERROR_INVALID_USM_SIZE);
 
-  ur_result_t Result = UR_RESULT_SUCCESS;
-  try {
-    ScopedContext Active(hContext->getDevice());
-    Result = UR_CHECK_ERROR(hipHostMalloc(ppMem, size));
-  } catch (ur_result_t Error) {
-    return Error;
+  if (!hPool) {
+    return USMHostAllocImpl(ppMem, hContext, nullptr, size, alignment);
   }
 
-  if (Result == UR_RESULT_SUCCESS) {
-    assert((!pUSMDesc || pUSMDesc->align == 0 ||
-            reinterpret_cast<std::uintptr_t>(*ppMem) % pUSMDesc->align == 0));
-    hContext->addUSMMapping(*ppMem, size);
+  auto UMFPool = hPool->HostMemPool.get();
```

Comment:
In fact, the same goes for the case where the handle/pointer is known and we follow into `umfPoolAlignedMalloc`. Perhaps we could generalise over those 3 (`urUSM[Host|Device|Shared]Alloc`) functions to avoid code duplication?

Reply:
I've created a new helper function for this.